Proposal: close() Method for the sidePanel API (#837)
hober/web-extensions: proposals/sidepanel-close-api.md

# New close() Method for the sidePanel API

**Summary**

> Proposal to introduce a programmatic close method for the side panel API.

Reviewer: Add a new line after `**Summary**` to prevent the sentance that follows from rendering on the same line. This rendering issue is visible [here](https://github.com/hharshas/webextensions/blob/255679a4b0073b51fe47cfe675d743b4f8c4da79/proposals/sidepanel-close-api.md).
```suggestion
**Summary**

Proposal to introduce a programmatic close method for the side panel API.
```

**Document Metadata**

**Author:** [hharshas](https://github.com/hharshas)

**Sponsoring Browser:** Chromium

> **Contributors:** [hharshas](https://github.com/hharshas) (
>   [Summer Of Code](https://summerofcode.withgoogle.com/)), mentor: oliverdunk, solomonkinard

Reviewer: Placing the opening parentheses (`(`) on one line and the content inside it on the next inserts a space between the the parentheses and its content. (Visible [here](https://github.com/hharshas/webextensions/blob/255679a4b0073b51fe47cfe675d743b4f8c4da79/proposals/sidepanel-close-api.md))

```suggestion
**Contributors:** [hharshas](https://github.com/hharshas) ([Summer Of
Code](https://summerofcode.withgoogle.com/)), mentor: oliverdunk, solomonkinard
```

**Created:** 2025-05-27

**Related Issues:** [#521](https://github.com/w3c/webextensions/issues/521),

>  [#chromium](https://issues.chromium.org/issues/403765214)

Reviewer: Minor: Trim leading space.

```suggestion
**Related Issues:** [#521](https://github.com/w3c/webextensions/issues/521),
[#chromium](https://issues.chromium.org/issues/403765214)
```

## Motivation

### Objective

What does this API enable?
- sidePanel.close() method will allow the sidebar to close programmatically.

> Why do we need it?
> - Currently, extensions cannot directly close their own sidebar from the
>  background script, 
> developers have to:
>   1. Send a message from a background script to the side panel.
>   2. Call window.close() from the side panel window on message delivery.

Reviewer: Fixes a rendering bug where "developers have to:" appeared as a paragraph separating the unordered list that precedes it from the ordered list that follows.

Also styles "window.close()" as monospace to signal this text is a snippet of code.

```suggestion
Why do we need it?
- Currently, extensions cannot directly close their own sidebar from the
  background script, developers have to:
  1. Send a message from a background script to the side panel.
  2. Call `window.close()` from the side panel window on message delivery.
```

This event aims to give developers an easier control.

#### Use Cases

Programmatic Control: 
- An extension could automatically close the panel after completing an action or
 when certain conditions are met. e.g. Closing the sidePanel after longer

>   inactivity.

Reviewer: Two minor issues with this block:

1. Several tech style guides recommend against using "e.g." in place of "for example" ([Apple](https://support.apple.com/guide/applestyleguide/a-apsg3acde405/1.0/web/1.0#apde9a26a57d4df4), [Google](https://developers.google.com/style/abbreviations#dont-use), [Microsoft](https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/e/eg)). While this group does not follow a specific style guide, it's a good practice to be aware of when writing technical content.

2. This line (and other parts of the document) use inconsistent spacing prefixes. This can cause issues as Markdown parsers do not necessarily handle single line breaks and varying indentation levels consistently. It's a bit long, but [this section](https://github.github.com/gfm/#motivation) of the GFM spec discusses some related concerns.

```suggestion
- An extension could automatically close the panel after completing an action or
  when certain conditions are met. For example, closing the sidePanel after longer
  inactivity.
```

### Known Consumers

[The Chromium bug](https://issues.chromium.org/issues/403765214) has a significant

>  amount of developer interest and the discussions on the issue shows the same.

Reviewer: Minor: Trim leading white space.

```suggestion
[The Chromium bug](https://issues.chromium.org/issues/403765214) has a significant
amount of developer interest and the discussions on the issue shows the same.
```

## Specification

### Schema

```typescript
namespace sidePanel {
  dictionary CloseOptions {
    // The window in which to close the side panel.
    long? windowId;

    // The tab for which to close the side panel. If specified,
    // only closes tab-specific side panels.
    long? tabId;

    // At least one of these must be provided, otherwise rejects with an error
  };

  interface Functions {
    // Closes the extension's side panel.
    // `options`: Specifies the context in which to close the side panel.
    // `callback`: Called when the side panel has been closed.
    static void close(
      CloseOptions options,
      optional VoidCallback callback);
  };
};
```

### Return value

- Type: Promise<void>

- Resolves when the panel has been closed. If the panel is 
already closed, still resolves successfully.

> ### Behavior
> - The operation will only close side panels that belong to the 
> calling extension (extension ID must match).
> - If the panel is already closed or does not exist in the given 
> context, the method does nothing.
> - If neither windowId nor tabId is provided, rejects with an error 
> - If windowId or tabId is invalid, rejects with an error.
> - If both windowId and tabId are provided, the method will verify 
> that the tab belongs to the specified window. If not, it rejects with an error.
> ### New Permissions
> N/A.
> 
> ### Manifest File Changes
> N/A

Reviewer: This suggestion addresses a few inconsistencies.

- Separate headers and contentent before/after them with a single blank line.
- Terminate all list items with period for consistency.
- Use inline monospace annotations to clearly signal which words are direct code references, such as an object's properties.
- Trim trailing whitespace.
- Normalize list item wrapping so the first character of a wrapped line aligns with the first character of content in the list item. 

```suggestion
### Behavior

- The operation will only close side panels that belong to the calling extension
  (extension ID must match).
- If the panel is already closed or does not exist in the given context, the
  method does nothing.
- If neither `windowId` nor `tabId` is provided, rejects with an error.
- If `windowId` or `tabId  is invalid, rejects with an error.
- If both `windowId` and `tabId` are provided, the method will verify that the
  tab belongs to the specified window. If not, it rejects with an error.

### New Permissions

N/A

### Manifest File Changes

N/A
```

## Security and Privacy

### Exposed Sensitive Data

This API does not expose any sensitive data.

### Abuse Mitigations

The API is limited to closing only the extension's own side panel.

### Additional Security Considerations

N/A

## Alternatives

### Existing Workarounds

* Message passing between background and panel contexts.

> * Calling window.close() from the panel context.

Reviewer: Style inline code snippet as monospace text.

```suggestion
* Calling `window.close()` from the panel context.
```

### Open Web API

Not applicable to the open web as it only complements the existing `open()` 
functionality.

## Implementation Notes

> N/A.

Reviewer: For consistency with the rest of the document, don't terminate a paragraph that only contains "N/A" with a period. 

```suggestion
N/A
```

## Future Work

This implementation will align with 
[#779](https://github.com/w3c/webextensions/pull/779), to ensure proper event

>  handling on panel closure. (Open for discussion)

Reviewer: Trim leading whitespace. 

```suggestion
This implementation will align with 
[#779](https://github.com/w3c/webextensions/pull/779), to ensure proper event
handling on panel closure. (Open for discussion)
```

**Summary**

Proposal to introduce a programmatic close method for the side panel API.

**Document Metadata**

**Author:** [hharshas](https://github.com/hharshas)

**Sponsoring Browser:** Chromium

**Contributors:** [hharshas](https://github.com/hharshas) ([Summer Of
Code](https://summerofcode.withgoogle.com/)), mentor: oliverdunk, solomonkinard

**Created:** 2025-05-27

**Related Issues:** [#521](https://github.com/w3c/webextensions/issues/521),
[#chromium](https://issues.chromium.org/issues/403765214)

## Motivation

### Objective

What does this API enable?
- `sidePanel.close()` method will allow the sidebar to close programmatically.

Why do we need it?
- Currently, extensions cannot directly close their own side panel from the
  background script, developers have to:
  1. Send a message from a background script to the side panel.
  2. Call `window.close()` from the side panel window on message delivery.

This event aims to give developers an easier control.

#### Use Cases

Programmatic Control:
- An extension could automatically close the panel after completing an action or
  when certain conditions are met. For example, closing the side panel after longer

>   inactivity.

Reviewer 2: I don't think inactivity is a compelling use case - could we think of another one? One that comes to mind is an extension that does some work / operates on a page and then closes when its work is finished.

namespace sidePanel {
  dictionary CloseOptions {
    // The window in which to close the side panel.
    long? windowId;

    // The tab for which to close the side panel. If specified,
    // only closes tab-specific side panels.
    long? tabId;

    // At least one of these must be provided, otherwise rejects with an error.
  };

  interface Functions {
    // Closes the extension's side panel.
    // `options`: Specifies the context in which to close the side panel.
    // `callback`: Called when the side panel has been closed.
    static void close(
      CloseOptions options,
      optional VoidCallback callback);
  };
};
```

### Return value

- Type: Promise<void>

- Resolves when the panel has been closed. If the panel is
  already closed, still resolves successfully.

### Behavior

- The operation will only close side panels that belong to the
  calling extension.
- If the panel is already closed or does not exist in the given
  context, the method does nothing.
- If neither `windowId` nor `tabId` is provided, rejects with an error.
- If `windowId` or `tabId` is invalid, rejects with an error.
- If both `windowId` and `tabId` are provided, the method will verify

>   that the tab belongs to the specified window. If not, it rejects with an error.

Reviewer 2: Adding a thread here to make sure we add a note on the behavior when there is a global side panel open and both `windowId` and `tabId` are provided.

  streamlining the user experience.

### Known Consumers

[The Chromium bug](https://issues.chromium.org/issues/403765214) has a significant
amount of developer interest and the discussions on the issue shows the same.

## Specification

### Schema

```typescript
namespace sidePanel {
  dictionary CloseOptions {
    // The window in which to close the side panel.
    long? windowId;

    // The tab for which to close the side panel. If specified,
    // only closes tab-specific side panels.
    long? tabId;

    // At least one of these must be provided, otherwise rejects with an error.
  };

  interface Functions {
    // Closes the extension's side panel.
    // `options`: Specifies the context in which to close the side panel.
    // `callback`: Called when the side panel has been closed.
    static void close(
      CloseOptions options,
      optional VoidCallback callback);
  };
};
```

### Return value

- Type: Promise<void>

- Resolves when the panel has been closed. If the panel is
  already closed, still resolves successfully.

### Behavior

- The operation will only close side panels that belong to the
  calling extension.

> - If the panel is already closed or does not exist in the given
>   context, the method does nothing.

Reviewer 3: ```suggestion
```

Move this to the two different scenarios, in which the method is called with a `tabId` or without a `tabId`

- If neither `windowId` nor `tabId` is provided, rejects with an error.
- If `windowId` or `tabId` is invalid, rejects with an error.
- If both `windowId` and `tabId` are provided, the method will verify
  that the tab belongs to the specified window. If not, it rejects with an error.

> - If both `windowId` and `tabId` are provided:

Reviewer 3: ```suggestion
- If `tabId` is provided:
```

Calling only with a tabId should be possible

>     1. If a contextual side panel is visible on the specified `tabId`, it will be closed.
>     2. If a global side panel is visible on the given `tabId`,
>        it will be closed for all tabs in that window, not just the specified tab.

Reviewer 3: ```suggestion
    1. If no contextual side panel is configured for the specified `tabId`, reject with an error.
    2. If the contextual side panel is currently not open, the method does nothing.
    3. If the contextual side panel is currently open, it will close the contextual side panel.
```

> 

Reviewer 3: ```suggestion
- If `tabId` is not provided:
    1. If no global side panel is currently open in the window, this method does nothing.
    2. If a global side panel is currently open in the window, close just the global side panel.
```
